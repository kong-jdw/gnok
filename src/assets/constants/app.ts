export const PUBLISHED = 'published'
export const UNPUBLISHED = 'unpublished'
export const IN_PROGRESS = 'in-progress'
export const CARDS_PER_PAGE = 12