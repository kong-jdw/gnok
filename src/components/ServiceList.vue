<script lang="ts" setup>
import ServiceCard from '@components/ServiceCard.vue'
import type { Service } from '@/composables/useServices'

interface Props {
  services: Service[],
  loading: boolean,
}

defineProps<Props>()
</script>

<template>
  <div v-if="loading">
    loading
  </div>

  <div
    v-else-if="services.length"
    class="card-list"
  >
    <ServiceCard
      v-for="service in services"
      :key="service.id"
      :service="service"
    />
  </div>

  <div
    v-else
    data-testid="no-results"
  >
    No services
  </div>
</template>

<style lang="scss" scoped>
@use '@css/media.scss';
@use '@css/spacing.scss';

.card-list {
  display: flex;
  flex-wrap: wrap;
  gap: spacing.$space-md;
}

.card-list > * {
  width: 100%;

  @include media.md {
    max-width: calc(50% - (spacing.$space-md / 2));
  }

  @include media.xl {
    max-width: calc(33.33% - (spacing.$space-md * 2 / 3));
  }
}
</style>
