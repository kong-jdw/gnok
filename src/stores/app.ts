import { defineStore } from 'pinia'

export const useAppStore = defineStore({
  id: 'app',

  state: () => ({
    serviceModalOpen: false,
  }),

  getters: {
    isCreateServiceModalOpen: (state) => state.serviceModalOpen,
  },

  actions: {
    toggleCreateServiceModal(): void {
      this.serviceModalOpen = !this.isCreateServiceModalOpen
    },
  },
})

