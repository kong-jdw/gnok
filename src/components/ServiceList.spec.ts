import { describe, it, expect, beforeEach } from 'vitest'
import { faker } from '@faker-js/faker'
import { mount } from '@vue/test-utils'
import ServiceList from './ServiceList.vue'
import servicesData from '../../mocks/services'
import { CARDS_PER_PAGE } from '@/assets/constants/app'

describe('ServiceCatalog', () => {
  const render = ({
    loading = false,
    error = false,
    services = servicesData,
  } = {}) => {
    return mount(ServiceList, {
      props: {
        loading,
        error,
        services,
      },
      shallow: true,
    })
  }

  beforeEach(() => {
  })

  it.each([
    ['empty', { services: [] }],
    ['loading', { loading: true }],
    ['error', { error: true }],
    ['normally', { services: servicesData.slice(0, 1) }],
  ])('renders %s', (kind: string, props) => {
    const wrapper = render(props)
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('does not have page navigation when only a couple services exist', () => {
    const wrapper = render({ services: servicesData.slice(0, 3) })
    const pageCount = wrapper.findTestId('pagination-controls')
    expect(pageCount.exists()).toBe(false)
  })

  it('has page navigation when more services exist than can be shown at once', () => {
    const services = Array(30).fill(undefined).map(() => ({
      ...servicesData[Math.floor(Math.random() * servicesData.length)],
      id: faker.string.uuid(), // ensure no :key collisions with limited mock data
    }))
    const wrapper = render({ services })
    const pageCount = wrapper.getTestId('pagination-controls')
    expect(pageCount.text()).toContain(`1 to ${CARDS_PER_PAGE} of 30 services`)
  })
})
