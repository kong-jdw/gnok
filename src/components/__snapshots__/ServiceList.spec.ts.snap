// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ServiceCatalog > renders empty 1`] = `
"<div data-v-9afa7982="" class="no-results" data-testid="no-results">
  <h3 data-v-9afa7982=""> No services found </h3>
  <anonymous-stub data-v-9afa7982="" icon="fa-solid:sad-cry"></anonymous-stub>
</div>"
`;

exports[`ServiceCatalog > renders error 1`] = `
"<div data-v-9afa7982="" class="error" data-testid="error">
  <h3 data-v-9afa7982=""> There was an error loading your services, please refresh and try again. </h3>
  <anonymous-stub data-v-9afa7982="" icon="hugeicons:sad-dizzy"></anonymous-stub>
</div>"
`;

exports[`ServiceCatalog > renders loading 1`] = `
"<div data-v-9afa7982="" class="loading" data-testid="loading">
  <loading-dots-stub data-v-9afa7982=""></loading-dots-stub>
</div>"
`;

exports[`ServiceCatalog > renders normally 1`] = `
"<div data-v-9afa7982="">
  <div data-v-9afa7982="" class="card-list">
    <service-card-stub data-v-9afa7982="" service="[object Object]"></service-card-stub>
  </div>
  <!--v-if-->
</div>"
`;
